devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
+
+`job_queue` is a module that allows jobs to be executed asynchronously. It queues jobs when there are no requests and queues requests when there are no jobs. When a job arrives and there are queued requests, it executes the least recently queued request with the job. When a request arrives and there are queued jobs, it executes the request with the least recently queued job.
+
+Consumers of `job_queue` must create a "`session`" before they can queue requests. A session can make two kinds of requests:
+
+- `get` requests
+- `pop` requests
+
+The consumer provides a `JOB_REQUEST_FUNCTION` that is executed when that request is executed. If the `JOB_REQUEST_FUNCTION` returns `true`, that job is considered completed, and it's  `JOB_COMPLETE_CALLBACK` is called. Once a job is completed, any new sessions that will be started will not see the completed job.
```

Comment:
>new sessions that will be started will not see the completed job [](http://example.com/codeflow?start=249&length=64)

Do current sessions all see the job even though it's completed?

Does every session see every job? Or is a job just processed by one session unless the processing fails?

Reply:
Let's say job X is pushed when there are two sessions active: session A and session B.

Session B calls `job_queue_session_pop` and the `JOB_REQUEST_FUNCTION` from this pop request returns `true` which triggers job X's `JOB_COMPLETE_CALLBACK`. 

Session A will still see the job when it calls `job_queue_session_pop`.

Every session sees every job that is submitted after `job_queue_session_begin` for that session was called. Jobs are processed by multiple sessions.